pkg/atomicwriter/atomic_writer.go

```diff
@@ -0,0 +1,109 @@
+package atomicwriter
+
+import (
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/cyberark/conjur-authn-k8s-client/pkg/log"
+	"github.com/cyberark/secrets-provider-for-k8s/pkg/log/messages"
+)
+
+const deleteTempFileRetryInterval = time.Second * 5
+const deleteTempFileMaxRetries = 5
+
+type atomicWriter struct {
+	path        string
+	permissions os.FileMode
+	tempFile    *os.File
+}
+
+// This package provides a simple atomic file writer which implements the
+// io.WriteCloser interface. This allows us to use AtomicWriter the way we
+// would use any other Writer, such as a Buffer. Additonally, this struct
+// takes the file path during construction, so the code which calls
+// `Write()` doesn't need to be concerned with the destination, just like
+// any other writer.
+func NewAtomicWriter(path string, permissions os.FileMode) (io.WriteCloser, error) {
+	dir, file := filepath.Split(path)
+	if dir == "" {
+		dir = "."
+	}
+
+	f, err := ioutil.TempFile(dir, file)
```

Comment:
@andytinkham I'm thinking it would be better to put this in a temp directory so it's not in the shared volume. Do you agree?

Reply:
Update: putting this in a temp folder results in a `invalid cross-device link` error when trying to call os.rename when the destination directory is on a different volume.